File: frontend/scripts/diagnose-missing-lines.py
#!/usr/bin/env python3
"""Diagnóstico geometría ACONDICIONAMIENTO — líneas faltantes."""
import json
import re
import subprocess
import sys
import urllib.request

XLSX = "/workspace/SEMANAS 2026.xlsx"
BASE = sys.argv[1] if len(sys.argv) > 1 else ""

WEEK_DAYS = ["lunes", "martes", "miercoles", "jueves", "viernes"]


def normalize_key(s: str) -> str:
    import unicodedata
    s = s.strip().lower()
    s = unicodedata.normalize("NFD", s)
    return "".join(c for c in s if unicodedata.category(c) != "Mn")


def row_text(row):
    return " ".join(c.strip() for c in row if c.strip()).lower()


def is_week_anchor(row):
    t = normalize_key(row_text(row))
    return all(d in t for d in WEEK_DAYS)


def extract_day_columns(row):
    m = {}
    for i, cell in enumerate(row):
        k = normalize_key(cell)
        if k == "lunes": m[i] = "Lunes"
        elif k == "martes": m[i] = "Martes"
        elif k in ("miercoles", "miércoles"): m[i] = "Miércoles"
        elif k == "jueves": m[i] = "Jueves"
        elif k == "viernes": m[i] = "Viernes"
    return m


def parse_line_cell(cell: str):
    n = normalize_key(cell)
    m = re.match(r"^linea\s*(\d+)\b", n)
    if m: return f"Línea {m.group(1)}"
    m = re.match(r"^l(\d+)$", n)
    if m: return f"Línea {m.group(1)}"
    m = re.match(r"^linea\s*n[°º]?\s*(\d+)", n)
    if m: return f"Línea {m.group(1)}"
    m = re.match(r"^premium\s*([ab])$", n)
    if m: return f"Premium {m.group(1).upper()}"
    return None


def detect_line_header(row):
    for cell in row:
        if not cell.strip(): continue
        line = parse_line_cell(cell.strip())
        if line: return line
    cells = [c.strip() for c in row if c.strip()]
    if 0 < len(cells) <= 4:
        joined = normalize_key(" ".join(cells))
        m = re.search(r"\blinea\s*(\d+)\b", joined)
        if m and len(cells) <= 2:
            return f"Línea {m.group(1)}"
    return None


def detect_sector(row):
    t = normalize_key(row_text(row))
    if "envasado consumo masivo" in t or "consumo masivo" in t:
        return "ENVASADO_MASIVO"
    if "envasado productos premiun" in t or "productos premiun" in t or "premium" in t:
        return "ENVASADO_PREMIUM"
    return None


def col_letter(n):
    s = ""
    while n > 0:
        n, r = divmod(n - 1, 26)
        s = chr(65 + r) + s
    return s


def scan_context(rows, target_row, target_col):
    ctx = {
        "sector_header": None, "sector_header_row": None,
        "line_header": None, "line_header_row": None,
        "active_sector": None, "active_line": None,
        "week_row": None, "day_label": None, "nearby": [],
    }
    day_columns = {}
    current_sector = None
    current_line = None

    for i, row in enumerate(rows):
        row_num = i + 1
        if is_week_anchor(row):
            day_columns = extract_day_columns(row)
            ctx["week_row"] = row_num
            current_line = None  # reset line on new week?
            continue

        sector = detect_sector(row)
        if sector:
            current_sector = sector
            line = detect_line_header(row)
            if line:
                current_line = line
            if row_num <= target_row:
                ctx["sector_header"] = " | ".join(c.strip() for c in row if c.strip())
                ctx["sector_header_row"] = row_num
                if line:
                    ctx["line_header"] = line
                    ctx["line_header_row"] = row_num
            continue

        line = detect_line_header(row)
        if line:
            current_line = line
            if row_num <= target_row:
                ctx["line_header"] = line
                ctx["line_header_row"] = row_num
            continue

        if row_num == target_row:
            for ci, day in day_columns.items():
                if ci + 1 == target_col:
                    ctx["day_label"] = day
            for ci, cell in enumerate(row):
                if cell.strip():
                    ctx["nearby"].append(f"{col_letter(ci+1)}{row_num}={cell.strip()[:40]}")
            for back in range(1, 8):
                prev = rows[i - back] if i >= back else None
                if not prev: break
                cells = [f"{col_letter(ci+1)}={c.strip()[:25]}" for ci, c in enumerate(prev) if c.strip()]
                if cells:
                    ctx["nearby"].insert(0, f"r{row_num-back}: {', '.join(cells[:5])}")

    ctx["active_sector"] = current_sector
    ctx["active_line"] = current_line
    return ctx


def parse_range(sr):
    m = re.match(r"^([^!]+)!(\d+):(\d+)$", sr or "")
    if not m: return None
    return {"sheet": m.group(1), "row": int(m.group(2)), "col": int(m.group(3))}


def fetch_api_items():
    if not BASE: return []
    import os
    bypass = os.environ.get("VERCEL_AUTOMATION_BYPASS_SECRET", "")
    items = []
    for sector in ("ENVASADO_MASIVO", "ENVASADO_PREMIUM"):
        req = urllib.request.Request(f"{BASE}/api/v1/work-items?sector={sector}")
        req.add_header("Accept", "application/json")
        if bypass:
            req.add_header("x-vercel-protection-bypass", bypass)
        with urllib.request.urlopen(req, timeout=120) as resp:
            data = json.loads(resp.read())
            items.extend(data.get("workItems", []))
    return items


def main():
    rows = json.loads(subprocess.check_output(["python3", "/workspace/frontend/scripts/load-acondicionamiento.py"]))
    items = fetch_api_items()
    print(f"Rows: {len(rows)}, API items: {len(items)}")

    missing = [i for i in items if i.get("sector") in ("ENVASADO_MASIVO", "ENVASADO_PREMIUM") and not i.get("line")]
    masivo = [i for i in missing if i["sector"] == "ENVASADO_MASIVO"]
    premium = [i for i in missing if i["sector"] == "ENVASADO_PREMIUM"]
    print(f"Sin línea: {len(missing)} (Masivo={len(masivo)}, Premium={len(premium)})")

    def pick(arr, n):
        if len(arr) <= n: return arr
        step = max(1, len(arr) // n)
        return [arr[min(i * step, len(arr)-1)] for i in range(n)]

    selected = pick(masivo, 12) + pick(premium, 8)

    print("\n| # | Sector | sourceRange | sectorHdr | lineHdr | activeLine | producto |")
    print("|---|--------|-------------|-----------|---------|------------|----------|")
    for i, item in enumerate(selected[:25]):
        loc = parse_range(item.get("sourceRange", ""))
        ctx = scan_context(rows, loc["row"], loc["col"]) if loc else {}
        print(
            f"| {i+1} | {item['sector']} | {item.get('sourceRange','?')} "
            f"| r{ctx.get('sector_header_row','?')} {(ctx.get('sector_header') or '—')[:25]} "
            f"| r{ctx.get('line_header_row','?')} {ctx.get('line_header') or '—'} "
            f"| {ctx.get('active_line') or 'null'} "
            f"| {(item.get('product') or item.get('plannedProduct') or '—')[:22]} |"
        )

    # Scan all rows for line header variants in sheet
    variants = {}
    for i, row in enumerate(rows):
        for ci, cell in enumerate(row):
            c = cell.strip()
            if not c: continue
            nk = normalize_key(c)
            if "linea" in nk or re.match(r"^l\d+$", nk) or "premium" in nk:
                key = c[:50]
                variants.setdefault(key, []).append(f"r{i+1}{col_letter(ci+1)}")

    print("\n## Variantes de línea en sheet (muestra)")
    for k, locs in list(variants.items())[:30]:
        print(f"  '{k}' → {', '.join(locs[:4])}{'...' if len(locs)>4 else ''}")

    # Count weeks
    weeks = [i+1 for i, row in enumerate(rows) if is_week_anchor(row)]
    print(f"\n## Semanas detectadas en filas: {weeks[:15]}{'...' if len(weeks)>15 else ''} ({len(weeks)} total)")

    print("\n## Detalle casos 1-3")
    for i, item in enumerate(selected[:3]):
        loc = parse_range(item.get("sourceRange", ""))
        ctx = scan_context(rows, loc["row"], loc["col"]) if loc else {}
        print(json.dumps({"case": i+1, "item": {k: item.get(k) for k in ("sector","product","sourceRange","weekLabel","dayLabel")}, "ctx": ctx}, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: frontend/scripts/diagnose-missing-lines.mjs
#!/usr/bin/env node
/**
 * Diagnóstico: WorkItems de envasado sin línea.
 * Usa API preview si BASE_URL está configurado, o parser local sobre SEMANAS 2026.xlsx.
 */
import { readFileSync, existsSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { createWorkItemRegistry } from "../src/lib/domain/work-item/work-item-registry.js";
import { workItemAssembler } from "../src/lib/domain/work-item/work-item-assembler.js";
import { projectDomainWorkItems } from "../src/lib/domain/work-item/work-item-projector.js";
import { parsePlannerTab } from "../src/lib/parsers/planner/planner-parser.js";
import {
  detectLineHeader,
  detectPackagingSectorHeader,
  extractDayColumns,
  isWeekAnchorRow,
  normalizeCellText,
  normalizeKey,
} from "../src/lib/parsers/planner/planner-utils.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, "../..");
const XLSX_PATH = join(ROOT, "SEMANAS 2026.xlsx");

const BASE = (process.env.BASE_URL ?? "").replace(/\/$/, "");
const BYPASS = process.env.VERCEL_AUTOMATION_BYPASS_SECRET?.trim() ?? "";

function headers(extra = {}) {
  const h = { Accept: "application/json", ...extra };
  if (BYPASS) h["x-vercel-protection-bypass"] = BYPASS;
  return h;
}

async function fetchWorkItems(sector) {
  const res = await fetch(`${BASE}/api/v1/work-items?sector=${sector}`, { headers: headers() });
  const body = await res.json();
  return body.workItems ?? [];
}

function parseSourceRange(range) {
  if (!range) return null;
  const m = range.match(/^([^!]+)!(\d+):(\d+)$/);
  if (!m) return { raw: range };
  return { sheet: m[1], row: Number(m[2]), col: Number(m[3]) };
}

async function loadLocalAcondicionamientoRows() {
  const { execSync } = await import("node:child_process");
  const script = `
import json, openpyxl
wb = openpyxl.load_workbook("${XLSX_PATH.replace(/\\/g, "/")}", data_only=True, read_only=True)
ws = wb["ACONDICIONAMIENTO"]
rows = []
for row in ws.iter_rows(values_only=True):
    rows.append([str(c) if c is not None else "" for c in row])
print(json.dumps(rows))
`;
  const out = execSync(`python3 -c ${JSON.stringify(script)}`, { maxBuffer: 50 * 1024 * 1024 });
  return JSON.parse(out.toString());
}

function colLetter(n) {
  let s = "";
  while (n > 0) {
    const r = (n - 1) % 26;
    s = String.fromCharCode(65 + r) + s;
    n = Math.floor((n - 1) / 26);
  }
  return s;
}

function buildRowContext(rows, targetRow, targetCol) {
  const ctx = {
    sectorHeader: null,
    sectorHeaderRow: null,
    lineHeader: null,
    lineHeaderRow: null,
    weekLabel: null,
    dayLabel: null,
    nearby: [],
  };

  let currentSector = null;
  let currentLine = null;
  let dayColumns = new Map();

  for (let i = 0; i < rows.length; i++) {
    const row = rows[i] ?? [];
    const rowNum = i + 1;

    if (isWeekAnchorRow(row)) {
      dayColumns = extractDayColumns(row);
      ctx.weekLabel = `fila ${rowNum}`;
      continue;
    }

    const packaging = detectPackagingSectorHeader(row);
    if (packaging) {
      currentSector = packaging;
      const line = detectLineHeader(row);
      if (line) {
        currentLine = line;
        if (rowNum <= targetRow) {
          ctx.sectorHeader = row.filter((c) => c.trim()).join(" | ");
          ctx.sectorHeaderRow = rowNum;
          ctx.lineHeader = line;
          ctx.lineHeaderRow = rowNum;
        }
      } else if (rowNum <= targetRow) {
        ctx.sectorHeader = row.filter((c) => c.trim()).join(" | ");
        ctx.sectorHeaderRow = rowNum;
      }
      continue;
    }

    const line = detectLineHeader(row);
    if (line) {
      currentLine = line;
      if (rowNum <= targetRow) {
        ctx.lineHeader = line;
        ctx.lineHeaderRow = rowNum;
      }
      continue;
    }

    if (rowNum === targetRow) {
      for (const [colIdx, day] of dayColumns) {
        if (colIdx + 1 === targetCol) ctx.dayLabel = day;
      }
      for (let c = 0; c < row.length; c++) {
        const cell = normalizeCellText(row[c] ?? "");
        if (cell) ctx.nearby.push(`${colLetter(c + 1)}${rowNum}=${cell.slice(0, 40)}`);
      }
      for (let back = 1; back <= 5; back++) {
        const prev = rows[i - back];
        if (!prev) break;
        const cells = prev.map((c, ci) => (c.trim() ? `${colLetter(ci + 1)}=${c.trim().slice(0, 30)}` : null)).filter(Boolean);
        if (cells.length) ctx.nearby.unshift(`fila${rowNum - back}: ${cells.slice(0, 4).join(", ")}`);
      }
    }
  }

  ctx.activeSectorAtRow = currentSector;
  ctx.activeLineAtRow = currentLine;
  return ctx;
}

function diagnoseLocal(rows, items) {
  const missing = items.filter(
    (i) =>
      (i.sector === "ENVASADO_MASIVO" || i.sector === "ENVASADO_PREMIUM") && !i.line
  );

  const samples = [];
  const masivo = missing.filter((i) => i.sector === "ENVASADO_MASIVO");
  const premium = missing.filter((i) => i.sector === "ENVASADO_PREMIUM");

  const pick = (arr, n) => {
    if (arr.length <= n) return arr;
    const step = Math.floor(arr.length / n);
    return Array.from({ length: n }, (_, i) => arr[Math.min(i * step, arr.length - 1)]);
  };

  const selected = [
    ...pick(masivo, 12),
    ...pick(premium, 8),
  ].slice(0, 25);

  for (const item of selected) {
    const loc = parseSourceRange(item.sourceRange);
    const ctx = loc?.row
      ? buildRowContext(rows, loc.row, loc.col ?? 1)
      : { nearby: [] };

    samples.push({
      sector: item.sector,
      product: item.product ?? item.plannedProduct ?? "—",
      sourceRange: item.sourceRange,
      tab: item.sourceSheet ?? loc?.sheet,
      row: loc?.row,
      col: loc?.col ? colLetter(loc.col) : null,
      weekLabel: item.weekLabel ?? ctx.weekLabel,
      sectorHeader: ctx.sectorHeader,
      sectorHeaderRow: ctx.sectorHeaderRow,
      lineHeader: ctx.lineHeader,
      lineHeaderRow: ctx.lineHeaderRow,
      activeLineAtRow: ctx.activeLineAtRow,
      dayLabel: item.dayLabel ?? ctx.dayLabel,
      nearby: ctx.nearby?.slice(0, 6),
      parserWouldDetectLine: ctx.lineHeader ?? ctx.activeLineAtRow,
    });
  }

  return { missing: missing.length, masivo: masivo.length, premium: premium.length, samples };
}

async function main() {
  let items = [];
  let rows = null;

  if (BASE) {
    const masivo = await fetchWorkItems("ENVASADO_MASIVO");
    const premium = await fetchWorkItems("ENVASADO_PREMIUM");
    items = [...masivo, ...premium];
    console.log(`API: ${masivo.length} Masivo + ${premium.length} Premium`);
  }

  if (existsSync(XLSX_PATH)) {
    console.log(`Parseando local: ${XLSX_PATH}`);
    rows = await loadLocalAcondicionamientoRows();
    const registry = createWorkItemRegistry();
    parsePlannerTab({
      fileId: "local-semanas",
      tab: "ACONDICIONAMIENTO",
      rows,
      registry,
      assembler: workItemAssembler,
    });
    const localItems = projectDomainWorkItems(registry.list());
    if (!items.length) items = localItems;
    console.log(`Local parser: ${localItems.length} items ACONDICIONAMIENTO`);
  }

  if (!rows && existsSync(XLSX_PATH)) {
    rows = await loadLocalAcondicionamientoRows();
  }

  const diag = diagnoseLocal(rows ?? [], items);

  console.log(`\nSin línea: ${diag.missing} (Masivo=${diag.masivo}, Premium=${diag.premium})\n`);
  console.log("| # | Sector | sourceRange | sectorHeader | lineHeader | activeLine | producto |");
  console.log("|---|--------|-------------|--------------|------------|------------|----------|");

  diag.samples.forEach((s, i) => {
    console.log(
      `| ${i + 1} | ${s.sector} | ${s.sourceRange} | r${s.sectorHeaderRow ?? "?"} ${(s.sectorHeader ?? "—").slice(0, 30)} | r${s.lineHeaderRow ?? "?"} ${s.lineHeader ?? "—"} | ${s.activeLineAtRow ?? "null"} | ${(s.product ?? "—").slice(0, 25)} |`
    );
  });

  console.log("\n## Detalle muestra (primeros 5)\n");
  diag.samples.slice(0, 5).forEach((s, i) => {
    console.log(`### Caso ${i + 1}`);
    console.log(JSON.stringify(s, null, 2));
  });
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
